core/server/proxy/src/main/java/alluxio/proxy/s3/CompleteMultipartUploadHandler.java

```diff
@@ -224,85 +233,37 @@ public CompleteMultipartUploadTask(FileSystem metaFs, FileSystem userFs,
-        uploadedParts.sort(new S3RestUtils.URIStatusNameComparator());
-        if (uploadedParts.size() < request.getParts().size()) {
-          throw new S3Exception(objectPath, S3ErrorCode.INVALID_PART);
-        }
-        Map<Integer, URIStatus> uploadedPartsMap = uploadedParts.stream().collect(Collectors.toMap(
-            status -> Integer.parseInt(status.getName()),
-            status -> status
-        ));
-        int lastPartNum = request.getParts().get(request.getParts().size() - 1).getPartNumber();
-        for (CompleteMultipartUploadRequest.Part part : request.getParts()) {
-          if (!uploadedPartsMap.containsKey(part.getPartNumber())) {
-            throw new S3Exception(objectPath, S3ErrorCode.INVALID_PART);
-          }
-          if (part.getPartNumber() != lastPartNum // size requirement not applicable to last part
-              && uploadedPartsMap.get(part.getPartNumber()).getLength() < Configuration.getBytes(
-                  PropertyKey.PROXY_S3_COMPLETE_MULTIPART_UPLOAD_MIN_PART_SIZE)) {
-            throw new S3Exception(objectPath, S3ErrorCode.ENTITY_TOO_SMALL);
-          }
-        }
+        List<URIStatus> uploadedParts = validateParts(request, objectPath, multipartTemporaryDir);
 
-        CreateFilePOptions.Builder optionsBuilder = CreateFilePOptions.newBuilder()
-            .setRecursive(true)
-            .setMode(PMode.newBuilder()
-                .setOwnerBits(Bits.ALL)
-                .setGroupBits(Bits.ALL)
-                .setOtherBits(Bits.NONE).build())
-            .setWriteType(S3RestUtils.getS3WriteType());
-        // Copy Tagging xAttr if it exists
-        if (metaStatus.getXAttr().containsKey(S3Constants.TAGGING_XATTR_KEY)) {
-          optionsBuilder.putXattr(S3Constants.TAGGING_XATTR_KEY,
-              ByteString.copyFrom(metaStatus.getXAttr().get(S3Constants.TAGGING_XATTR_KEY)));
-        }
-        // Copy Content-Type Header xAttr if it exists
-        if (metaStatus.getXAttr().containsKey(S3Constants.CONTENT_TYPE_XATTR_KEY)) {
-          optionsBuilder.putXattr(S3Constants.CONTENT_TYPE_XATTR_KEY,
-              ByteString.copyFrom(metaStatus.getXAttr().get(S3Constants.CONTENT_TYPE_XATTR_KEY)));
-        }
-        AlluxioURI objectUri = new AlluxioURI(objectPath);
-        try {
-          S3RestUtils.deleteExistObject(mUserFs, objectUri);
-        } catch (IOException | AlluxioException e) {
-          throw S3RestUtils.toObjectS3Exception(e, objectUri.getPath());
-        }
-        // (re)create the merged object
+        // (re)create the merged object to a temporary object path
         LOG.debug("CompleteMultipartUploadTask (bucket: {}, object: {}, uploadId: {}) "
             + "combining {} parts...", mBucket, mObject, mUploadId, uploadedParts.size());
-        FileOutStream os = mUserFs.createFile(objectUri, optionsBuilder.build());
+        CreateFilePOptions createFileOption = prepareForCreateTempFile(metaStatus);
```

Comment:
Do you think we should check whether the final multipart object exists and its upload ID is same as the current upload ID before creating the temporary object?
If so, we don't need to complete the multipart again.

Reply:
if an mp upload obj has been completed and then before the deletemptemp dir call has happened, we will be proceeding to the create temp file here, think this is a very small gap, most case would be we exception out at time of checking the mp temp folder side. think the gap is quite small,  actually planning to add some metrics later (not part of this PR) track the latency of each call, and see how huge that gap maybe, and add an additional check if gap is large.

Reply:
I just think we need to read all parts and then write all data to the temp. And then we find we have already had the target object. it's not necessary. What do you think?

Reply:
yeah we will proceed to create temp obj if we found the temp dir exists in checkStatusesForUploadId step, if someone has completed the target obj, it would start deleting the temp dir in removePartsDirAndMPMetaFile call and it would make us get an exception from checkStatusesForUploadId and not proceed with the create temp obj, the case u r describing is a very small gap : when someone has completed the target obj creation and before doing removePartsDirAndMPMetaFile, and I come in and pass the checkStatusesForUploadId and start createFile, do u see what i'm saying ?

Reply:
okay, that makes sense